offscreen/scripts/offscreen.ts

```diff
@@ -9,41 +11,63 @@ import initLedger from './ledger';
 import initTrezor from './trezor';
 import initLattice from './lattice';
 
-initLedger();
-initTrezor();
-initLattice();
-
 /**
  * Initialize a post message stream with the parent window that is initialized
  * in the metamask-controller (background/serivce worker) process. This will be
  * utilized by snaps for communication with snaps running in the offscreen
  * document.
  */
-const parentStream = new BrowserRuntimePostMessageStream({
-  name: 'child',
-  target: 'parent',
-});
-
-ProxySnapExecutor.initialize(parentStream, './snaps/index.html');
-
-if (process.env.IN_TEST) {
-  chrome.runtime.onMessage.addListener((message) => {
-    if (
-      message &&
-      typeof message === 'object' &&
-      message.event === OffscreenCommunicationEvents.metamaskBackgroundReady &&
-      message.target === OffscreenCommunicationTarget.extension
-    ) {
-      window.document?.documentElement?.classList?.add('controller-loaded');
-    }
+function initializePostMessageStream() {
+  const parentStream = new BrowserRuntimePostMessageStream({
+    name: 'child',
+    target: 'parent',
   });
+
+  ProxySnapExecutor.initialize(parentStream, './snaps/index.html');
 }
 
-chrome.runtime.sendMessage({
-  target: OffscreenCommunicationTarget.extensionMain,
-  isBooted: true,
+/**
+ * Initialize the ledger, trezor, and lattice keyring connections, and the
+ * post message stream for the Snaps environment.
+ */
+async function init(): Promise<void> {
+  initializePostMessageStream();
+  initTrezor();
+  initLattice();
 
-  // This message is being sent from the Offscreen Document to the Service Worker.
-  // The Service Worker has no way to query `navigator.webdriver`, so we send it here.
-  webdriverPresent: navigator.webdriver === true,
+  try {
+    const ledgerInitTimeout = new Promise((_, reject) => {
+      setTimeout(() => {
+        reject(new Error('Ledger initialization timed out'));
+      }, OFFSCREEN_LEDGER_INIT_TIMEOUT);
+    });
+    await Promise.race([initLedger(), ledgerInitTimeout]);
+  } catch (error) {
+    console.error('Ledger initialization failed:', error);
+  }
+}
+
+init().then(() => {
+  if (process.env.IN_TEST) {
+    chrome.runtime.onMessage.addListener((message) => {
+      if (
+        message &&
+        isObject(message) &&
+        message.event ===
+          OffscreenCommunicationEvents.metamaskBackgroundReady &&
+        message.target === OffscreenCommunicationTarget.extension
+      ) {
+        window.document?.documentElement?.classList?.add('controller-loaded');
+      }
+    });
+  }
+
+  chrome.runtime.sendMessage({
```

Comment:
Hmm, this still makes extension bootup and every other service using the offscreen document dependent on the Ledger iframe. You also would not be able to unlock before 4s has passed or the Ledger iframe loads.

Wondering if we should be emitting a separate event for Ledger and awaiting that in the keyring somewhere not on init as the longer-term solution.

Reply:
I think we should evaluate how useful it is to always add a Ledger keyring to the user's keychain on unlock, even if the user never interacted with a ledger device. If we establish that this is not needed, then as you suggested the best thing to do would be to initialize Ledger only when needed (e.g. when the Keyring is explicitly added by a user), so that we don't have to wait for it before unlocking